source/Ship.cpp

```diff
@@ -3266,39 +3276,75 @@ bool Ship::CanBeCarried() const
 
 
 
-bool Ship::Carry(const shared_ptr<Ship> &ship)
+bool Ship::Carry(const shared_ptr<Ship> &ship, vector<Visual> *visuals)
 {
 	if(!ship || !ship->CanBeCarried())
 		return false;
 	
-	// Check only for the category that we are interested in.
-	const string &category = ship->attributes.Category();
+	Bay *dockBay = ship->reservedBay;
+	if(!dockBay)
+	{
+		// The ship to carry has no bay assigned to it. Find one.
+
+		// Check only for the category that we are interested in.
+		const string &category = ship->attributes.Category();
+		for(Bay &bay : bays)
+			if(bay.category == category && bay.free)
+			{
+				dockBay = &bay;
+				break;
+			}
+
+		if(!dockBay)
+			return false;
+	}
 	
-	for(Bay &bay : bays)
-		if((bay.category == category) && !bay.ship)
-		{
-			bay.ship = ship;
-			ship->SetSystem(nullptr);
-			ship->SetPlanet(nullptr);
-			ship->SetTargetSystem(nullptr);
-			ship->SetTargetStellar(nullptr);
-			ship->SetParent(shared_from_this());
-			ship->isThrusting = false;
-			ship->isReversing = false;
-			ship->isSteering = false;
-			ship->commands.Clear();
-			// If this fighter collected anything in space, try to store it
-			// (unless this is a player-owned ship).
-			if(!isYours && cargo.Free() && !ship->Cargo().IsEmpty())
-				ship->Cargo().TransferAll(cargo);
-			// Return unused fuel to the carrier, for any launching fighter that needs it.
-			ship->TransferFuel(ship->fuel, this);
+	dockBay->ship = ship;
+	dockBay->free = false;
+	ship->reservedBay = dockBay;
+	ship->SetSystem(nullptr);
+	ship->SetPlanet(nullptr);
+	ship->SetTargetSystem(nullptr);
+	ship->SetTargetStellar(nullptr);
+	ship->SetParent(shared_from_this());
+	ship->isThrusting = false;
+	ship->isReversing = false;
+	ship->isSteering = false;
+	ship->commands.Clear();
+	// If this fighter collected anything in space, try to store it
+	// (unless this is a player-owned ship).
+	if(!isYours && cargo.Free() && !ship->Cargo().IsEmpty())
+		ship->Cargo().TransferAll(cargo);
+	// Return unused fuel to the carrier, for any launching fighter that needs it.
+	ship->TransferFuel(ship->fuel, this);
 			
-			// Update the cached mass of the mothership.
-			carriedMass += ship->Mass();
-			return true;
+	// Update the cached mass of the mothership.
+	carriedMass += ship->Mass();
+
+	// Apply any retrieve effects if applicable (for example, no effect should
+	// be generated when launching from a planet).
+	if(visuals)
+		for(const Effect *effect : dockBay->retrieveEffects)
+			visuals->emplace_back(*effect, ship->Position(), ship->Velocity(), ship->Facing());
+	return true;
+}
+
+
+
+// Track the given escort as a carry in one bay while in flight.
+void Ship::CarryInFlight(const shared_ptr<Ship> &escort)
```

Comment:
`Carry` is when the carry boards the ship. Specifically, it does things like settings planets and system to `nullptr`, storing cargo inside its parent, ...

We need to have a way for carries to tell the parent that they want a bay. Previously this wasn't needed carries that need a bay would just 1) as the parent if there are free bays and 2) then board. This is not possible anymore since the carry (and the ship) need to now beforehand which bay they will dock too. Hence this function.

Reply:
Sounds like `ReserveBay()` is a better name for this function then (since the carried ship is not actually carried after this function, but only seems to have reserved a bay).

Reply:
I agree! Thanks for the suggestion.